library/constant_time.c

```diff
@@ -47,16 +47,58 @@
 
 #include <string.h>
 
+/*
+ * Define MBEDTLS_EFFICIENT_UNALIGNED_VOLATILE_ACCESS where assembly is present to
+ * perform fast unaligned access to volatile data.
+ *
+ * This is needed because mbedtls_get_unaligned_uintXX etc don't support volatile
+ * memory accesses.
+ *
+ * Some of these definitions could be moved into alignment.h but for now they are
+ * only used here.
+ */
+#if defined(MBEDTLS_EFFICIENT_UNALIGNED_ACCESS) && defined(MBEDTLS_HAVE_ASM)
+#if defined(__arm__) || defined(__thumb__) || defined(__thumb2__) || defined(__aarch64__)
+#define MBEDTLS_EFFICIENT_UNALIGNED_VOLATILE_ACCESS
+#endif
+#endif
+
+#if defined(MBEDTLS_EFFICIENT_UNALIGNED_VOLATILE_ACCESS)
+static inline uint32_t mbedtls_get_unaligned_volatile_uint32( volatile const unsigned char *p )
+{
+    /* This is UB, even where it's safe:
+    *    return *((volatile uint32_t*)p);
+    * so instead the same thing is expressed in assembly below.
+    */
+    uint32_t r;
+#if defined(__arm__) || defined(__thumb__) || defined(__thumb2__)
+    asm( "ldr %0, [%1]" : "=r"(r) : "r"(p) : );
+#elif defined(__aarch64__)
+    asm( "ldr %w0, [%1]" : "=r"(r) : "r"(p) : );
+#endif
+    return r;
+}
+#endif /* MBEDTLS_EFFICIENT_UNALIGNED_VOLATILE_ACCESS */
+
 int mbedtls_ct_memcmp( const void *a,
                        const void *b,
                        size_t n )
 {
-    size_t i;
+    size_t i = 0;
     volatile const unsigned char *A = (volatile const unsigned char *) a;
     volatile const unsigned char *B = (volatile const unsigned char *) b;
-    volatile unsigned char diff = 0;
+    volatile uint32_t diff = 0;
```

Comment:
So how about something like

```
    /* `A` and `B` are volatile because otherwise, once `diff` gets to -1
     * (which happens ra[pidly if A and B are random and independent),
     * no further reading of either is necessary. */
    volatile const unsigned char *A = (volatile const unsigned char *) a;
    volatile const unsigned char *B = (volatile const unsigned char *) b;

    /* `diff` is volatile to ensure that the compiler always does the xor
    // (as it has to read and write it every iteration). */
    volatile unsigned char diff = 0;
```